### PR TITLE
Properly fix dagrun update state endpoint

### LINKED ISSUE #18363: http patched an in-progress dagRun to state=failed, but the change was clobbered by a subsequent TI state change.
### Apache Airflow version

2.2.0b1 (beta snapshot)

### Operating System

debian buster

### Versions of Apache Airflow Providers

n/a

### Deployment

Astronomer

### Deployment details

`astro dev start` image: [quay.io/astronomer/ap-airflow-dev:2.2.0-buster-43897](https://quay.io/repository/astronomer/ap-airflow-dev?tab=tags)

### What happened

I patched a dagRun's state to `failed` while it was running.

```
curl -i -X PATCH "http://localhost:8080/api/v1/dags/steady_task_stream/dagRuns/scheduled__2021-09-18T00:00:00+00:00"   -H 'Content-Type: application/json' --user 'admin:admin' \
 -d '{ "state": "failed" }'
```

For a brief moment, I saw my change reflected in the UI.  Then auto-refresh toggled itself to disabled.  When I re-enabled it, the dag_run the state: "running".

new_api

Presumably this happened because the new API **only** affects the dagRun.  The tasks keep running, so they updated dagRun state and overwrite the patched value.

### What you expected to happen

The UI will not let you set a dagrun's state without also setting TI states. Here's what it looks like for "failed":
ui

Notice that it forced me to fail a task instance.  This is what prevents the dag from continuing onward and clobbering my patched value.  It's similar when you set it to "success", except every child task gets set to "success".

At the very least, *if* the API lets me patch the dagrun state *then* it should also lets me patch the TI states. This way an API user can patch both at once and prevent the clobber.  (`/api/v1/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}` doesn't yet support the `patch` verb, so I can't currently do this).

Better still would be for the API to make it easy for me to do what the UI does.  For instance, patching a dagrun with:
```
{ "state": "failed" }
```
Should return code 409 and a message like:

> To patch steady_task_stream.scheduled__2021-09-18T00:00:00+00:00.state to success, you must also make the following changes: { "task_1":"success", "task_2":"success", ... }.  Supply "update_tasks": "true" to do this.

And updating it with:
```
{ "state": "failed",
  "update_tasks": "true"}
```
Should succeed and provide feedback about which state changes occurred.

### How to reproduce

Any dag will do here, but here's the one I used: 
 https://gist.github.com/MatrixManAtYrService/654827111dc190407a3c81008da6ee16

Be sure to run it in an airflow that has https://github.com/apache/airflow/pull/17839, which introduced the patch functionality that makes it possible to reach this bug.

- Unpause the dag
- Make note of the execution date
- Delete the dag and wait for it to repopulate in the UI (you may need to refresh).
- Prepare this command in your terminal, you may need to tweak the dag name and execution date to match your scenario:
```
curl -i -X PATCH "http://localhost:8080/api/v1/dags/steady_task_stream/dagRuns/scheduled__2021-09-18T00:00:00+00:00"   -H 'Content-Type: application/json' --user 'admin:admin' \
 -d '{ "state": "failed" }'
```
- Unpause the dag again
- Before it completes, run the command to patch its state to "failed"

Note that as soon as the next task completes, your patched state has been overwritten and is now "running" or maybe "success"

### Anything else

@ephraimbuddy, @kaxil mentioned that you might be interested in this.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ephraimbuddy**
The dagrun update state endpoint was recently added but is not working
as expected. This PR fixes it to work exactly like the UI mark dagrun state API

Closes: https://github.com/apache/airflow/issues/18363

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by kaxil** [comment]
Some tests are failing:


```
==== API mysql: 2 failures ====

tests/api_connexion/endpoints/test_dag_run_endpoint.py::TestPatchSetDagRunState::test_should_respond_200[failed]: AssertionError: assert 'running' == 'failed'
  - failed
  + running
tests/api_connexion/endpoints/test_dag_run_endpoint.py::TestPatchSetDagRunState::test_should_respond_200[success]: AssertionError: assert 'running' == 'success'
  - success
  + running
```

**Comment by ephraimbuddy** [comment]
> Some tests are failing:
> 
> ```
> ==== API mysql: 2 failures ====
> 
> tests/api_connexion/endpoints/test_dag_run_endpoint.py::TestPatchSetDagRunState::test_should_respond_200[failed]: AssertionError: assert 'running' == 'failed'
>   - failed
>   + running
> tests/api_connexion/endpoints/test_dag_run_endpoint.py::TestPatchSetDagRunState::test_should_respond_200[success]: AssertionError: assert 'running' == 'success'
>   - success
>   + running
> ```

I have been trying to figure why this fails when I run the whole endpoints but doesn't fail when I run only the module.
Running this fails: `pytest tests/api_connexion/endpoints/`
but running this doesn't fail: `pytest tests/api_connexion/endpoints/test_dag_run_endpoint.py`

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 1]
```diff
@@ -21,6 +21,10 @@
 from sqlalchemy import or_
 
 from airflow._vendor.connexion import NoContent
+from airflow.api.common.experimental.mark_tasks import (
+    set_dag_run_state_to_failed,
+    set_dag_run_state_to_success,
+)
```
Comment (kaxil) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:27]: Can we do without using experimental API ?

Comment (ephraimbuddy) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:27]: Hi Kaxil, I think the utility functions inside` common/experimental` package are not yet deprecated but the experimental REST API is? The experimental API lives in `api/client` package.  I tried not using these functions but the change was big. Now I'm suggesting we use it but move the useful util functions out of the `/experimental` or rename the `/experimental` to api_utils or something appropriate?

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:27]: There are actually quite a few places depending on `airflow.api.common.experimental` in the web UI and the new API. We should do a more comprehensive refactoring to move those out of `experimental` instead, so I feel it should be OK-ish to just referring them right here.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 2]
```diff
@@ -302,6 +306,10 @@ def update_dag_run_state(dag_id: str, dag_run_id: str, session) -> dict:
         raise BadRequest(detail=str(err))
 
     state = post_body['state']
-    dag_run.set_state(state=DagRunState(state))
-    session.merge(dag_run)
+    dag = current_app.dag_bag.get_dag(dag_id)
+    if state == State.SUCCESS:
+        set_dag_run_state_to_success(dag, dag_run.execution_date, commit=True)
+    else:
+        set_dag_run_state_to_failed(dag, dag_run.execution_date, commit=True)
```
Comment (SamWheating) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py:313]:
```suggestion
    elif state == State.FAILED:
    	set_dag_run_state_to_failed(dag, dag_run.execution_date, commit=True)
    else:
    	raise BadRequest(f"Bad value for `state`: {state}")
```

If the `state` field in the post body is anything other than `"success"`, this will cause the DAGRun to be marked as failed.

For example, if someone tries to update a DAGRun with `{state: succes}` it will mark the DAGRun as failed due to the spelling error.

I think its a lot safer to do nothing / return an error in the event of an unexpected or input.

Comment (ephraimbuddy) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:313]: We have a check in the open api documentation that won’t allow any other state apart from `success` and `failed` State to get to this endpoint

Comment (ephraimbuddy) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:313]: You can check the test cases for this endpoint

Comment (SamWheating) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py:313]: My mistake, I didn't know about - Thanks for the explanation.

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py [hunk 3]
```diff
@@ -1162,16 +1162,19 @@ def test_should_raises_403_unauthorized(self):
...
+    def test_should_respond_200(self, state, session):
         dag_id = "TEST_DAG_ID"
         dag_run_id = 'TEST_DAG_RUN_ID'
-        with dag_maker(dag_id):
-            DummyOperator(task_id='task_id')
-        dr = dag_maker.create_dagrun(run_id=dag_run_id)
+        date = datetime(2021, 1, 1)
+        dag = DAG(dag_id, schedule_interval=None, start_date=date)
+        task = DummyOperator(task_id='task_id', dag=dag)
+        dag.tasks.append(task)
```
Comment (uranusjr) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]: Why is this needed?

Comment (ephraimbuddy) [reply tests/api_connexion/endpoints/test_dag_run_endpoint.py]: Removed. Added when I couldn't find why dag was not having tasks.

**Review by uranusjr — APPROVED** [review]
